### ISSUE TITLE
[e2e failure] [sig-network] Networking Granular Checks: Services [Slow] should function for client IP based session affinity: http

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/kubernetes#54571]
/priority critical-urgent
/sig network

This test case has been failing for a while and affects a number of jobs: [triage report](https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=%5C%5Bsig-network%5C%5D%20Networking%20Granular%20Checks%3A%20Services%20%5C%5BSlow%5C%5D%20should%20function%20for%20client%20IP%20based%20session%20affinity%3A%20http#87b694524eb439d53a6f)

This is affecting multiple jobs on the release-master-blocking dashboard, and prevents us from cutting 1.9.0-alpha.2 (kubernetes/sig-release#22). Is there work ongoing to bring this job back to green?

Possibly related:
- #53835
- #53568

[triage cluster 679a79064b46aae4d4b3](https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=%5C%5Bsig-network%5C%5D%20Networking%20Granular%20Checks%3A%20Services%20%5C%5BSlow%5C%5D%20should%20function%20for%20client%20IP%20based%20session%20affinity%3A%20http#679a79064b46aae4d4b3)
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/networking.go:225
Oct 18 09:36:14.886: Expect endpoints: map[], got: map[netserver-0:{}]
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/networking.go:222
```

**Comment by spiffxp** [comment]
@kubernetes/sig-network-test-failures

**Comment by MrHohn** [comment]
Fix https://github.com/kubernetes/kubernetes/pull/53760 is opened but waiting for approval.

**Comment by MrHohn** [comment]
Humm..Test is passing on gce-slow but still very flaky on gce-alpha after #53760 went in:
- slow: https://k8s-testgrid.appspot.com/google-gce#gci-gce-slow&sort-by-failures=
- alpha: https://k8s-testgrid.appspot.com/google-gce#gci-gce-alpha-features&sort-by-failures=

cc @m1093782566

**Comment by m1093782566** [comment]
I believe proxy mode in gce-alpha is iptables?

**Comment by m1093782566** [comment]
These test cases are about **ClientIP** based session affinity. From the log, it seems that we have successfully updated service session affinity type to **ClientIP**. However, the endpoints we retrieved are different. I can only suspect the clients which requests service in different tries are different. Possibly I can refactor the test case to make sure the client IPs in different tries are the same.

**Comment by MrHohn** [comment]
> BTW, I want to know the difference between gce-slow and gce-alpha. I am wondering why gce-slow passed but gce-alpha failed?

One difference I know of is we enable all alpha features in gce-alpha (including kube-proxy as daemonset, pod priority and others). Ref the env file we have:
https://github.com/kubernetes/test-infra/blob/master/jobs/env/ci-kubernetes-e2e-gci-gce-alpha-features.env
 
Existing alpha features are listed in https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go.

> I believe proxy mode in gce-alpha is iptables?

Yep, that sounds correct.

**Comment by spiffxp** [comment]
FWIW even though it looks like [triage cluster 679a79064b46aae4d4b3](https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=%5C%5Bsig-network%5C%5D%20Networking%20Granular%20Checks%3A%20Services%20%5C%5BSlow%5C%5D%20should%20function%20for%20client%20IP%20based%20session%20affinity%3A%20http#679a79064b46aae4d4b3) shows the same text, the actual failure from recent jobs now looks like
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/networking.go:204
Oct 27 15:24:12.545: Expect endpoints: map[netserver-1:{}], got: map[netserver-2:{}]
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/networking.go:227
```
This appears to be flaking more heavily (or getting more runs) in [gci-gke-alpha-features](https://k8s-testgrid.appspot.com/google-gke#gci-gke-alpha-features&sort-by-flakiness=&exclude-non-failed-tests=) as well

**Comment by MrHohn** [comment]
> I can only suspect the clients which requests service in different tries are different.

Not sure how this can happen, specifically from the code:
https://github.com/kubernetes/kubernetes/blob/4c8539cece2f0a6e6974b30d00c7341e10320bc5/test/e2e/framework/networking_utils.go#L243
We always send request from `config.HostTestContainerPod` and it never changes during test.

**Comment by spiffxp** [comment]
/remove-priority critical-urgent
/priority important-soon
ci-kubernetes-e2e-gci-gke-alpha-features is the only remaining failing job, which is not on release-master-blocking

Seems like kubernetes/test-infra#5095 didn't catch this job and a followup PR would be appropriate

/assign @danehans @MrHohn
since they took care of this failure for the gce job

**Comment by MrHohn** [comment]
Send out another fix https://github.com/kubernetes/kubernetes/pull/55122.

**Comment by spiffxp** [comment]
/status approved-for-milestone
tracking this for https://github.com/kubernetes/sig-release/issues/27

**Comment by MrHohn** [comment]
Tests are passing now:
- https://k8s-testgrid.appspot.com/sig-network#gci-gce-slow
- https://k8s-testgrid.appspot.com/sig-network#gci-gce-alpha-features

Closing this issue.
/close
